### PR TITLE
[SPARK-16307] [ML] Add test to verify the predicted variances of a DT on toy data

### PR DESCRIPTION
**Pull request by MechCoder**
## What changes were proposed in this pull request?

The current tests assumes that `impurity.calculate()` returns the variance correctly. It should be better to make the tests independent of this assumption. In other words verify that the variance computed equals the variance computed manually on a small tree.
## How was this patch tested?

The patch is a test....

### CODE REVIEW

**Comment by MechCoder** [comment]
@yanboliang Could you have a look?

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 1]
```diff
@@ -96,6 +108,15 @@ class DecisionTreeRegressorSuite
       assert(variance === expectedVariance,
         s"Expected variance $expectedVariance but got $variance.")
     }
+
+    val toyDF = TreeTests.setMetadata(toyData, Map.empty[Int, Int], 0)
+    dt.setMaxDepth(1)
+      .setMaxBins(6)
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Not sure why we need to set maxBins here?

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 2]
```diff
@@ -96,6 +108,15 @@ class DecisionTreeRegressorSuite
...
+      .setSeed(0)
+    val expectVariances = dt.fit(toyDF).transform(toyDF).select("variance").collect().map {
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: `expectedVariances` and `trueVariances` are mixed up here. Expected should be the theoretical value computed below. Also, it would be good to leave a comment explaining where those expected values came from.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 3]
```diff
@@ -96,6 +108,15 @@ class DecisionTreeRegressorSuite
...
+      case Row(variance: Double) => variance }
+    val trueVariances = Array(0.667, 0.667, 0.667, 2.667, 2.667, 2.667)
+    trueVariances.zip(expectVariances).foreach(x => x._1 ~== x._2 absTol 1e-3)
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Although this technically works, it is less confusing if use `assert` and unpack the tuple. Like

```scala
...foreach { case (actual, expected) =>
  assert(actual ~== expected absTol 1e-3)
}
```

**Comment by sethah** [comment]
@MechCoder Thanks for adding this! I think it's a good test to protect against silent failures in the future. I just left a few small comments.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 1 above)
Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Because there are 6 datapoints, and I want each datapoint to be a split.

Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: I verified and my intuition was correct. I get this warning for the default setting:

 WARN DecisionTreeMetadata: DecisionTree reducing maxBins from 32 to 6 (= number of training instances)

**Comment by MechCoder** [comment]
@sethah Thank you for your comments. I have addressed them. Do you have anything else?

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 1 above)
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Ok, not sure it needs to be set since the test is fine without it. The warning is more just helpful info logging. I'm fine to leave it since it doesn't matter either way.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 4]
```diff
@@ -96,6 +108,22 @@ class DecisionTreeRegressorSuite
       assert(variance === expectedVariance,
         s"Expected variance $expectedVariance but got $variance.")
     }
+
+    val toyDF = TreeTests.setMetadata(toyData, Map.empty[Int, Int], 0)
+    dt.setMaxDepth(1)
+      .setMaxBins(6)
+      .setSeed(0)
+    val calculatedVariances = dt.fit(toyDF).transform(toyDF).select("variance").collect().map {
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Can we use `.select(dt.getVarianceCol)` instead? Also, the closing brace should go on the next line, like:

```scala
.map {
  case Row(...)
}
```

**Comment by sethah** [comment]
@MechCoder I'm not sure how I feel about leaving the original test, which will fail if another impurity is added and IMO isn't really correct. Thoughts?

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 1 above)
Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: "Explicit is better than implicit" ;)

**Comment by MechCoder** [comment]
I'm slightly in favour of keeping the original test because the impurity is set to "variance" explicitly by the `setImpurity` method, so it's a safe assumption that the `calculate` method returns the variance.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 1 above)
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: I'd like to remove the explicit setting since the default value(32) meets your needs. We want to make Jenkins logging clean and reduce the number of warnings if possible.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 5]
```diff
@@ -36,10 +37,21 @@ class DecisionTreeRegressorSuite
...
   private var categoricalDataPointsRDD: RDD[LabeledPoint] = _
 
+  private var toyData: RDD[LabeledPoint] = _
+
   override def beforeAll() {
     super.beforeAll()
+
     categoricalDataPointsRDD =
       sc.parallelize(OldDecisionTreeSuite.generateCategoricalDataPoints().map(_.asML))
+    toyData = sc.parallelize(Seq(
```
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Move ```toyData``` to ```TreeTests```. You can refer [Feature importance with toy data](https://github.com/apache/spark/blob/master/mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala#L108).

**Comment by yanboliang** [comment]
@MechCoder I left some inline comments. Thanks for adding this test case.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 1 above)
Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: If you would like to reduce the number of warnings, then this should be kept as is (unless I am misunderstanding something)

Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: You are right, please ignore my last comments.

**Comment by MechCoder** [comment]
@yanboliang Thanks! Addressed your comments. Let me know if there is anything else.

**Comment by MechCoder** [comment]
OK, that should be it. I removed all the unused variables and imports.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 6]
```diff
@@ -96,6 +97,25 @@ class DecisionTreeRegressorSuite
...
         s"Expected variance $expectedVariance but got $variance.")
     }
+
+    val varianceData: RDD[LabeledPoint] = TreeTests.varianceData(sc)
+    val varianceDF = TreeTests.setMetadata(varianceData, Map.empty[Int, Int], 0)
+    dt.setMaxDepth(1)
+      .setMaxBins(6)
+      .setSeed(0)
+    val transformVarDF = dt.fit(varianceDF).transform(varianceDF)
+    val calculatedVariances = transformVarDF.select(dt.getVarianceCol).collect().map {
```
Comment (jaceklaskowski) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: I wonder if `toDS` worked here and you'd have `variance` "simpler".

Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: You mean after the collect?

It fails with `is not a member of Seq[org.apache.spark.sql.Row]`

Comment (jaceklaskowski) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Before `collect` so you work with type-safe `Dataset` (not a `DataFrame` which is `Dataset[Row]`)

**Comment by yanboliang** [comment]
LGTM. Merged into master. Thanks!

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 6 above)
Comment (yanboliang) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: @jaceklaskowski I think it's better to do such improvement in a separate PR which will involve lots of similar changes for all test suites? Thanks!

**Comment by MechCoder** [comment]
@yanboliang Would appreciate it if you could look at https://github.com/apache/spark/pull/13650

**Comment by MechCoder** [comment]
Thanks @sethah @yanboliang for the reviews!!

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 6 above)
Comment (MechCoder) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: Ah, I see. Thanks for the note. I wasn't familiar with the Dataset API till now.
